pkg/iac/terraform/block.go

```diff
@@ -576,3 +596,122 @@ func (b *Block) IsNil() bool {
 func (b *Block) IsNotNil() bool {
 	return !b.IsNil()
 }
+
+func (b *Block) ExpandBlock() error {
+	var (
+		expanded []*Block
+		errs     []error
+	)
+
+	for _, child := range b.childBlocks {
+		if child.Type() == "dynamic" {
+			blocks, err := child.expandDynamic()
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			expanded = append(expanded, blocks...)
+		}
+	}
+
+	for _, block := range expanded {
+		b.injectBlock(block)
+	}
+
+	return errors.Join(errs...)
+}
+
+func (b *Block) expandDynamic() ([]*Block, error) {
+	if b.IsExpanded() || b.Type() != "dynamic" {
+		return nil, nil
+	}
+
+	realBlockType := b.TypeLabel()
+	if realBlockType == "" {
+		return nil, errors.New("dynamic block must have 1 label")
+	}
+
+	forEachVal, err := b.validateForEach()
+	if err != nil {
+		return nil, fmt.Errorf("invalid for-each in %s block: %w", b.FullLocalName(), err)
+	}
+
+	var (
+		expanded []*Block
+		errs     []error
+	)
+
+	forEachVal.ForEachElement(func(key, val cty.Value) (stop bool) {
+		if val.IsNull() {
+			return
+		}
+
+		iteratorName, err := b.iteratorName(realBlockType)
+		if err != nil {
+			errs = append(errs, err)
+			return
+		}
+
+		forEachCtx := b.childContext()
+		obj := cty.ObjectVal(map[string]cty.Value{
+			"key":   key,
+			"value": val,
+		})
+		forEachCtx.Set(obj, iteratorName)
+
+		if content := b.GetBlock("content"); content != nil {
+			inherited := content.inherit(forEachCtx)
+			inherited.hclBlock.Labels = []string{}
+			inherited.hclBlock.Type = realBlockType
+			if err := inherited.ExpandBlock(); err != nil {
+				errs = append(errs, err)
+				return
+			}
+			expanded = append(expanded, inherited)
+		}
+		return
+	})
+
+	if len(expanded) > 0 {
+		b.markExpanded()
+	}
+
+	return expanded, errors.Join(errs...)
+}
+
+func (b *Block) validateForEach() (cty.Value, error) {
+	forEachAttr := b.GetAttribute("for_each")
+	if forEachAttr == nil {
+		return cty.NilVal, errors.New("for_each attribute required")
+	}
+
+	forEachVal := forEachAttr.Value()
+
+	if !forEachVal.CanIterateElements() {
+		return cty.NilVal, fmt.Errorf("cannot use a %s value in for_each. An iterable collection is required", forEachVal.GoString())
+	}
+
+	if forEachVal.IsNull() || !forEachVal.IsKnown() {
+		return cty.NilVal, errors.New("cannot use a null or unknown value in for_each")
+	}
```

Comment:
In my coverage this seems to be untested so I decided to investigate. Would this ever be reached since we already check for `!forEachVal.CanIterateElements()` prior to this? If you have a sample input, it'd be nice to understand.

Reply:
Yes, this check is unreachable, I removed it. https://github.com/aquasecurity/trivy/pull/7612/commits/4d3de2af0bf8805357edb7854c4d07e5b626334f